File: app/src/main/java/com/pacmac/devinfo/cellular/CellInfoFutureTask.java
package com.pacmac.devinfo.cellular;

import android.annotation.SuppressLint;
import android.os.Build;
import android.telephony.CellInfo;
import android.telephony.TelephonyManager;

import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executor;
import java.util.concurrent.FutureTask;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class CellInfoFutureTask extends FutureTask {

    private final static int TIMEOUT = 500;  // seconds

    private List<CellInfo> cellInfo = null;

    CellInfoFutureTask() {
        super(() -> null);
    }

    void run(List<CellInfo> cellInfo) {
        this.cellInfo = cellInfo;
        super.run();
    }

    List<CellInfo> getAllCellInfoBlocking() throws ExecutionException, InterruptedException, TimeoutException {
        super.get(TIMEOUT, TimeUnit.MILLISECONDS);
        return cellInfo;
    }

    @SuppressLint("MissingPermission")
    public static List<CellInfo> getAllCellInfoBlocking(TelephonyManager telephonyManager) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            try {
                final CellInfoFutureTask future = new CellInfoFutureTask();

                telephonyManager.requestCellInfoUpdate(new Executor() {
                    @Override
                    public void execute(Runnable cellInfoCallbackCallable) {
                        cellInfoCallbackCallable.run();
                    }
                }, new TelephonyManager.CellInfoCallback() {
                    @Override
                    public void onCellInfo(List<CellInfo> cellInfo) {
                        future.run(cellInfo);
                    }
                });
                return future.getAllCellInfoBlocking();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return null;
    }
}

File: app/src/main/java/com/pacmac/devinfo/cellular/CellInfoFutureTask.kt
package com.pacmac.devinfo.cellular

import android.annotation.SuppressLint
import android.os.Build
import android.telephony.CellInfo
import android.telephony.TelephonyManager
import android.util.Log
import java.util.concurrent.FutureTask
import java.util.concurrent.TimeUnit

internal class CellInfoFutureTask : FutureTask<Void>({ null }) {

    private var cellInfo: List<CellInfo>? = null

    fun run(cellInfo: List<CellInfo>) {
        this.cellInfo = cellInfo
        super.run()
    }

    @Throws(Exception::class)
    fun getAllCellInfoBlocking(): List<CellInfo>? {
        super.get(TIMEOUT_MS, TimeUnit.MILLISECONDS)
        return cellInfo
    }

    companion object {
        private const val TIMEOUT_MS = 500L

        @SuppressLint("MissingPermission")
        fun getAllCellInfoBlocking(telephonyManager: TelephonyManager): List<CellInfo>? {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return null
            return try {
                val future = CellInfoFutureTask()
                telephonyManager.requestCellInfoUpdate(
                    { it.run() },
                    object : TelephonyManager.CellInfoCallback() {
                        override fun onCellInfo(cellInfo: List<CellInfo>) {
                            future.run(cellInfo)
                        }
                    }
                )
                future.getAllCellInfoBlocking()
            } catch (e: Exception) {
                Log.e("CellInfoFutureTask", "Failed to get cell info", e)
                null
            }
        }
    }
}
